presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/AddExchanges.java

```diff
@@ -246,8 +246,17 @@ public PlanWithProperties visitAggregation(AggregationNode node, PreferredProper
                         child.getProperties());
             }
             else if (!child.getProperties().isStreamPartitionedOn(partitioningRequirement) && !child.getProperties().isNodePartitionedOn(partitioningRequirement)) {
+                checkState(preferredProperties.getGlobalProperties().isPresent() && preferredProperties.getGlobalProperties().get().getPartitioningProperties().isPresent());
```

Comment:
You should modify `partitioningRequirement` so that it contains merged partitioning requirement.
Otherwise this code would insert `partitionedExchange` even if data is already partitioning by merged requirement.
For example, `partitioningRequirement={a,b}`, `mergedPartitioningPreferences={a}`, `isNodePartitionedOn={a}`.

Here and similar in other methods.